### ISSUE TITLE
Conflicted Mapping causes problems during shard initialization

### ISSUE THREAD

**Issue by pickypg** [elastic/elasticsearch#17567]
This issue reproduces in ES 2.3.

```http
POST /my-index/_bulk
{"index":{"_type":"type1"}}
{"title":"I am just a string"}
{"index":{"_type":"type2"}}
{"title":{"field":"I am a string in an object"}}

POST /my-index/_close
POST /my-index/_open
```

After opening the index, you'll get something like this in the logs:

```
[2016-04-06 10:24:20,677][WARN ][cluster.action.shard     ] [Anthropomorpho] [my-index][0] received shard failed for target shard [[my-index][0], node[dC8bCuUMRHWqd0dmyqxgwg], [P], v[88], s[INITIALIZING], a[id=HdFXs5PdQI6iT4G5HFyF3w], unassigned_info[[reason=ALLOCATION_FAILED], at[2016-04-06T14:24:20.505Z], details[failed to update mappings, failure IllegalArgumentException[Field [title] is defined as a field in mapping [type1] but this name is already used for an object in other types]]]], indexUUID [LiIeIKE-RvSNnfaBUi9NWw], message [failed to update mappings], failure [IllegalArgumentException[Field [title] is defined as a field in mapping [type1] but this name is already used for an object in other types]]
java.lang.IllegalArgumentException: Field [title] is defined as a field in mapping [type1] but this name is already used for an object in other types
	at org.elasticsearch.index.mapper.MapperService.checkObjectsCompatibility(MapperService.java:473)
	at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:336)
	at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:289)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.processMapping(IndicesClusterStateService.java:387)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyMappings(IndicesClusterStateService.java:348)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.clusterChanged(IndicesClusterStateService.java:164)
	at org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:610)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:772)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
```

Relates to #15243

**Comment by clintongormley** [comment]
Closed by #17568

**Comment by damienalexandre** [comment]
I got this bug on production today, with mapping creation based on templates (logstash :heart:).

I had a field which was a string in 99% of my logs, and another "type", only getting event once an hour, with this field as an object - I didn't notice, as everything was working smoothly for weeks. Until we decided to install Marvel :neckbeard: 

Marvel is fine, but we restarded the cluster to start the agent... our 70+ shards were not able to start anymore! IllegalArgumentException on all shards, about this field not mapped properly.

I had to close all the indexes, no more logs history :sob: and since 2.0 removed the possibility to delete a Type, no possibility to clean up those bad indexes, I'm stuck with gigabyte of data I can't use because shard won't boot. Anyway.

Glad this is now fixed, and waiting for ES 2.3.2 :neckbeard: :sparkling_heart:
